docs/architecture/ErrorReporting.md

```diff
@@ -0,0 +1,122 @@
+# Handling and reporting errors
+
+    WARNING: This document describes future plans for error handling.
+    The UI library does not (yet) implement error handling as described here.
+
+This document describes how errors are handled within the ACS UI library and how they are reported to the surrounding application.
+
+## Error sources
+
+There are two categories of errors, and they are handled differently.
+
+1. ACS backend errors: Errors returned by the REST calls to ACS backend services.
+
+   1. These can be transient or permanent (special case: permission error).
+
+2. UI errors: Programming errors in the UI library / errors returned from other browser APIs
+
+## ACS backend errors
+
+Handling ACS backend errors is challenging when using either the composites or the pure components with the react bindings provided by the ACS UI library:
+
+* ACS backend API calls are asynchronous
+* ACS backend API calls are often initiated outside of the React component lifecycle methods (e.g on interaction handler callbacks or `useEffect` blocks).
+
+But in both cases, all ACS backend API calls are mediated by the (chat and calling) stateful clients. Thus, the stateful client tees errors to a special field in the state:
+
+```ts
+interface CallClientState {
+    // ... other fields ...
+    errors: ErrorState[];
+}
+
+interface ErrorState {
+    /**
+     * Locally generated unique ID for error.
+     * Allows for de-duplicating similar errors from repeated API calls.
+     */
+    id: string;
+    sdkClientMethod: 'CallClient.createCallAgent' | 'CallClient.getDeviceManager'; // ... and many more.
```

Comment:
do we define all of these in the stateful client projects?

Reply:
Yes.

Undecided which is worse: Duplicating the list of methods in this literal type, or just allowing `string` `sdkClientMethod`.

Since the errors are reported and used asynchronous to the method that failed, we need _some_ way to know what method failed.

Reply:
Update doc uses an object with fields keyed by the method names. Even more reason to use literal types now.